runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingService.java

```diff
@@ -0,0 +1,292 @@
+    return "artifact_" + Hashing.sha256().hashString(artifactMetadata.getName(), CHARSET)
+        .toString();
+  }
+
+  private StagingSessionToken decodeStagingSessionToken(String stagingSessionToken)
+      throws IOException {
+    try {
+      return MAPPER.readValue(stagingSessionToken, StagingSessionToken.class);
+    } catch (IOException e) {
+      try {
+        LOG.error(
+            "Unable to deserialize staging token {}. Expected format {}. Error {}",
+            stagingSessionToken, MAPPER.writeValueAsString(new StagingSessionToken()),
+            e.getMessage());
+      } catch (JsonProcessingException e1) {
+        LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+            StagingSessionToken.class);
+      }
+      throw e;
+    }
+  }
+
+  private static String encodeStagingSessionToken(StagingSessionToken stagingSessionToken)
+      throws Exception {
+    try {
+      return MAPPER.writeValueAsString(stagingSessionToken);
+    } catch (JsonProcessingException e) {
+      LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+          StagingSessionToken.class);
+      throw e;
+    }
+  }
+
+  private ResourceId getJobDirResourceId(String stagingSessionToken) throws IOException {
+    ResourceId baseResourceId;
+    StagingSessionToken parsedToken = decodeStagingSessionToken(stagingSessionToken);
+    try {
+      baseResourceId = FileSystems.matchSingleFileSpec(parsedToken.getBasePath())
+          .resourceId();
+    } catch (FileNotFoundException fne) {
+      // Create the base path
+      baseResourceId = FileSystems
+          .matchNewResource(parsedToken.getBasePath(), true /* isDirectory */);
+    }
+    // Using sessionId as the subDir to store artifacts and manifest.
+    return baseResourceId
+        .resolve(parsedToken.getSessionId(), StandardResolveOptions.RESOLVE_DIRECTORY);
+  }
+
+  private ResourceId getManifestFileResourceId(String stagingSessionToken) throws IOException {
+    return getJobDirResourceId(stagingSessionToken)
+        .resolve(MANIFEST, StandardResolveOptions.RESOLVE_FILE);
+  }
+
+  private ResourceId getArtifactDirResourceId(String stagingSessionToken) throws IOException {
+    return getJobDirResourceId(stagingSessionToken)
+        .resolve("artifacts", StandardResolveOptions.RESOLVE_DIRECTORY);
+  }
+
+  private class PutArtifactStreamObserver implements StreamObserver<PutArtifactRequest> {
+
+    private final StreamObserver<PutArtifactResponse> outboundObserver;
+    private PutArtifactMetadata metadata;
+    private ResourceId artifactId;
+    private WritableByteChannel artifactWritableByteChannel;
+
+    PutArtifactStreamObserver(StreamObserver<PutArtifactResponse> outboundObserver) {
+      this.outboundObserver = outboundObserver;
+    }
+
+    @Override
+    public void onNext(PutArtifactRequest putArtifactRequest) {
+      // Create the directory structure for storing artifacts in the first call.
```

Comment:
Hmm I wonder how are we going to handle parallel artifact upload streams in the future. Fine to leave as a TODO for now, I suspect it'd require some small API changes.

Reply:
It already handles it correctly and we have testcases :)

Each artifact upload opens a new putArtifact stream and will be managed independently.

Reply:
This should be fine for uploading multiple artifacts in parallel. I'm not sure uploading multiple chunks of the same artifact in parallel is worth the complexity.